contents/regex/Regex.md

```diff
@@ -0,0 +1,316 @@
+[Regex Examples](http://www.regular-expressions.info/examples.html)<br>
+[Regex Course](https://regexone.com/)
+
+# Advanced topics
+
+## The Regex Engine
+
+So, how does Regex work exactly?
+
+Regex works similarly to most brute force searches. A simplified Regex engine would look something like this:
+
+1. Start from the start of the text and the start of the regex.
+2. Check if the first character matches.
+3. If it does check the second character of both strings and so on.
+4. If any character doesn't match, start from the second character of the text.
+5. Return when a match is found, or if no match is found.
+
+For example, for the regex `light` against the sentence `He likes light lightbulbs.`:
+
+```
+He likes light lightbulbs.
+^
+light
+```
+
+First, the engine checks against the start of the string, `H` against `l`. This doesn't match, so the engine moves on to `e`, then ` `, and so on.
+
+```
+He likes light lightbulbs.
+   ^
+   light
+```
+
+Once it reaches the first `l`, it sees a match, consumes the `l` and starts comparing the rest of the string.
+
+```
+He likes light lightbulbs.
+     ^
+   light
+```
+
+However, when it reaches `k`, it finds a different character so it knows it's not a match. It then starts again from `i` (the letter after `l`).
+
+```
+He likes light lightbulbs.
+         ^
+         light
+```
+
+Finally, when it finds a match, it returns the match and terminates. Note that the second `light` in `lightbulbs` is ignored.
+
+Of course, more complex Regex functions will require more steps, but this will give you a rough understanding of the engine for now.
+
+## Greedy Operators
+
+The operators `{}`, `*` and `+` are called greedy operators because they try to match as many characters as possible. For example, for the regex `b.*` against the string `baaaaaa`:
+
+```
+baaaaaaaaa
+ ^
+ba
+```
+
+Although a suitable match (`ba`) has already been found, the engine sees that it can match against a longer string of `a`s. It will instead return the match `baaaaaaaaa`, matching the entire string.
+
+But what if a greedy operator can't find a match? A greedy operator will always consume as many characters as possible, but if the string that follows fails to match, the engine releases one character at a time until a match is found. This is called 'backtracking'. For example, consider the regex `c.*ed` against the string `contested` (Let `.....` represent `.*`).
+
+```
+contested
+        ^
+c........ed
+```
+
+Remember that `.*` matches any number of characters, and this includes the `ed` at the back! The entire string is matched by `c.*`, leaving no characters left for `ed`. This obviously doesn't match, so the engine backtracks one step:
+
+```
+contested
+        ^
+c.......ed
+```
+
+As we can see, `e` still doesn't match `d`, so we take another step back:
+
+```
+contested
+       ^
+c......ed
+```
+
+Finally, we get a match and see that both strings match.
+
+## Effect on performance
```

Comment:
This falls under greedy operators, right? If it is the case, should make the heading one level lower, i.e. `### Effect on performance`.